### DISCUSSION TITLE
Lazy loading providers that are defined inside configuration file

### DISCUSSION THREAD

**Discussion by afshinpir** [openssl/openssl#23731] [Q&A]
Hi All,
I'm trying to define an entry inside `openssl.cnf` and lazy load it later. So I wrote a config file like this:
```
...
[openssl_init]
providers = provider_sect

# List of providers to load
[provider_sect]
# default = default_sect
# The fips section name should match the section name inside the
# included fipsmodule.cnf.
base = base_sect
fips = fips_sect
myprovider= myprovider_sect

[myprovider_sect]
module = /data/myprovider.so
additional_info = test
...
```
As you see I have added entry without `activate=1` to my config file. Now, I was expecting that if I try to load config file and then use `OSSL_PROVIDER_load()`, I should be able to load my provider. So I wrote a code like this:
```
int func(OSSL_PROVIDER *provider, void *cbdata) {
        std::cout << "provider: " << OSSL_PROVIDER_get0_name(provider) << "\n";
        return 1;
}

int main(int argc, char** argv) {
    if (OSSL_LIB_CTX_load_config(NULL, "openssl.cnf") <= 0) {
        std::cout<< "FATAL: error loading configuration file\n";
        ERR_print_errors_fp(stderr);
        exit(1);
    }

    int ret = -1;
    OSSL_PROVIDER *ctx;
    ctx= OSSL_PROVIDER_load(NULL, "myprovider");
    if (ctx== NULL) {
        std::cout << "Failed to load Default provider\n";
        ERR_print_errors_fp(stderr);
        return -1;
    }
    std::cout << "load done\n";
    OSSL_PROVIDER_do_all(NULL, func, NULL);
    OSSL_PROVIDER_unload(ctx);
}
```
But this code does not work as expected. I traced openssl codes to some extend and I noticed that an entry for non-activated provider is added to internal structures, but it seems it is not searched and activated by calling `OSSL_PROVIDER_load()`. 

Is there anyway to achieve what I have in mind?

**Comment by mattcaswell** [comment]
> But this code does not work as expected

What did you expect to happen, and what actually happened?

>  noticed that an entry for non-activated provider is added to internal structures, but it seems it is not searched and activated by calling OSSL_PROVIDER_load()

It should be, and happens here:

https://github.com/openssl/openssl/blob/5677992679b38950c6a0c3775fd57378e1879ba5/crypto/provider_core.c#L542-L553

What version of OpenSSL are you using?

**Reply by afshinpir** [reply]
@mattcaswell 
Yes, I was expecting that too. But reality is a little different.
From my investigation I found out that one major problem is the fact that module path will not be set if we try to load module in this way. `template.path` should have module path here and it is need to be set after this line:
https://github.com/openssl/openssl/blob/5677992679b38950c6a0c3775fd57378e1879ba5/crypto/provider_core.c#L575
similar to what happens here:
https://github.com/openssl/openssl/blob/5677992679b38950c6a0c3775fd57378e1879ba5/crypto/provider_conf.c#L237-L238
but module path is not set. I try to add this myself and test things out, but module still does not load correctly.
I have yet to test if I'm missing something or not, but I always to load provider like this.
I'm using version 3.0.13 right now, but this code is similar in all version. And as you see in my config file, I'm using FIPS module and have disabled default provider.
